jobs/kubelet/templates/bin/ensure_kubelet_up_and_running.erb

```diff
@@ -2,11 +2,76 @@
 
 [ -z "$DEBUG" ] || set -x
 
+CERT_DIR=/var/lib/kubelet/pki
+
 kubectl="/var/vcap/packages/kubernetes/bin/kubectl --kubeconfig=/var/vcap/jobs/kubelet/config/kubeconfig"
+admin_kubectl="/var/vcap/packages/kubernetes/bin/kubectl --kubeconfig=/var/vcap/jobs/kubelet/config/admin-kubeconfig"
+
+wait_for_client_cert() {
+    until [ -e $CERT_DIR/kubelet-client.crt ]; do
+        sleep 2
+    done
+}
+
+wait_for_server_cert() {
+    # Fetch the nodename from the bootstrapped Kubelet certificate since it should be the same
+    # and will honor whatever the cloud provider overrides the name with
+    local nodename=$(openssl x509 -noout -subject -in $CERT_DIR/kubelet-client.crt | sed -n '/^subject/s/^.*CN=//p')
+
+    until $admin_kubectl get csr | grep -e ' Pending' | grep "$nodename"; do
+      sleep 2
+    done
+
+    local csr_name=$($admin_kubectl get csr | grep -e ' Pending' | grep "$nodename" | awk '{print $1}')
+    $admin_kubectl certificate approve $csr_name
```

Comment:
Rather than approving the csr and having the admin credentials on the kubelet, does it make more sense to have a job on the master that will continually poll for csr's from the kubelets and approve them?

Reply:
I think trying to run that CSR-approver on the Master runs into the same problem that has prevented Kubernetes from having an auto-approval for server certificates: there isn't a mechanism for the signing server to validate the CN and SANs in the cert. And so a Kubelet can request a cert for a domain that may be incorrect and have it automatically approved.

The admin credentials are promptly deleted once the Kubelet has finished its registration and I wasn't able to have them recreated without the startup scripts also being re-run (and thus deleting the admin creds once more). With this approach, we're auto-approving the CSR _specifically_ for the Kubelet that we're starting and then relinquishing control to the rotation mechanisms.

Reply:
I might be missing something and it might be helpful to have an actual discussion, but can we not get the information from bosh? I'm also not sure where the validation is happening. Can you point me to the logic that is validating the CN and SANs in the cert?

My only concern with the approach is that it feels unintuitive to have the approval happen on the worker because I wouldn't expect the worker to have the power to approve its own csr. We are deleting the admin credentials afterward which is fine, but ideally I think the worker wouldn't have them in the first place. That being said if we can't get the correct worker CN on the master via BOSH (probably via a link) then this is a no-op and we will probably have to go with your approach.